core/services/relay/evm/mercury/wsrpc/cache/cache.go

```diff
@@ -0,0 +1,401 @@
+	// LatestReportDeadline controls how long to wait for a response before
+	// retrying. Setting this to zero will wait indefinitely.
+	LatestReportDeadline time.Duration
+}
+
+func NewCache(client Client, cfg Config) Cache {
+	return newMemCache(client, cfg)
+}
+
+type cacheVal struct {
+	sync.RWMutex
+
+	fetching bool
+	fetchCh  chan (struct{})
+
+	val *pb.LatestReportResponse
+	err error
+
+	expiresAt time.Time
+}
+
+func (v *cacheVal) read() (*pb.LatestReportResponse, error) {
+	v.RLock()
+	defer v.RUnlock()
+	return v.val, v.err
+}
+
+// caller expected to hold lock
+func (v *cacheVal) initiateFetch() <-chan struct{} {
+	if v.fetching {
+		panic("cannot initiateFetch on cache val that is already fetching")
+	}
+	v.fetching = true
+	v.fetchCh = make(chan struct{})
+	return v.fetchCh
+}
+
+func (v *cacheVal) setError(err error) {
+	v.Lock()
+	defer v.Unlock()
+	v.err = err
+}
+
+func (v *cacheVal) completeFetch(val *pb.LatestReportResponse, err error, expiresAt time.Time) {
+	v.Lock()
+	defer v.Unlock()
+	if !v.fetching {
+		panic("can only completeFetch on cache val that is fetching")
+	}
+	v.val = val
+	v.err = err
+	if err == nil {
+		v.expiresAt = expiresAt
+	}
+	close(v.fetchCh)
+	v.fetchCh = nil
+	v.fetching = false
+}
+
+func (v *cacheVal) abandonFetch(err error) {
+	v.completeFetch(nil, err, time.Now())
+}
+
+func (v *cacheVal) waitForResult(ctx context.Context, chResult <-chan struct{}, chStop <-chan struct{}) (*pb.LatestReportResponse, error) {
+	select {
+	case <-ctx.Done():
+		_, err := v.read()
+		return nil, errors.Join(err, ctx.Err())
+	case <-chStop:
+		return nil, ErrStopped{}
+	case <-chResult:
+		return v.read()
+	}
+}
+
+// memCache stores values in memory
+// it will never return a stale value older than latestPriceTTL, instead
+// waiting for a successful fetch or caller context cancels, whichever comes
+// first
+type memCache struct {
+	services.StateMachine
+	lggr logger.Logger
+
+	client Client
+
+	latestPriceTTL       time.Duration
+	maxStaleAge          time.Duration
+	latestReportDeadline time.Duration
+
+	cache sync.Map
+
+	wg     sync.WaitGroup
+	chStop services.StopChan
+}
+
+func newMemCache(client Client, cfg Config) *memCache {
+	return &memCache{
+		services.StateMachine{},
+		cfg.Logger.Named("MercuryMemCache"),
+		client,
+		cfg.LatestReportTTL,
+		cfg.MaxStaleAge,
+		cfg.LatestReportDeadline,
+		sync.Map{},
+		sync.WaitGroup{},
+		make(chan (struct{})),
+	}
+}
+
+type ErrStopped struct{}
+
+func (e ErrStopped) Error() string {
+	return "memCache was stopped"
+}
+
+// LatestReport
+// NOTE: This will actually block on all types of errors, even non-timeouts.
+// Context should be set carefully and timed to be the maximum time we are
+// willing to wait for a result, the background thread will keep re-querying
+// until it gets one even on networking errors etc.
+func (m *memCache) LatestReport(ctx context.Context, req *pb.LatestReportRequest) (resp *pb.LatestReportResponse, err error) {
```

Comment:
Is this more complicated than it needs to be? I can see that it accounts for the cache wait case, but I'm wondering if you considered using a simpler strategy (cache aside, rather than read through). This might be a bit less efficient, but I think would significantly reduce all of the edge cases you need to handle, which feels like a win.

Reply:
Can you outline how the simpler strategy would work?

Reply:
Update: as above.